File: mahaam-api-go/feat/service/user_service.go
package service

import (
	"mahaam-api/feat/repo"
	"mahaam-api/infra/configs"
	"mahaam-api/infra/dbs"
	"mahaam-api/infra/emails"
	logs "mahaam-api/infra/log"
	"mahaam-api/infra/security"
	"slices"

	"github.com/google/uuid"
	"github.com/jmoiron/sqlx"
)

type UserService interface {
	Create(device Device) *CreatedUser
	SendMeOtp(email string) string
	VerifyOtp(meta Meta, email, sid, otp string) *VerifiedUser
	RefreshToken(meta Meta) *VerifiedUser
	UpdateName(userID UUID, name string) int64
	Logout(userID UUID, deviceId UUID) int64
	Delete(userID UUID, sid, otp string)
	GetDevices(userID UUID) []Device
	GetSuggestedEmails(userID UUID) []SuggestedEmail
	DeleteSuggestedEmail(userID UUID, suggestedEmailId UUID)
}

type userService struct {
	userRepo            repo.UserRepo
	deviceRepo          repo.DeviceRepo
	planRepo            repo.PlanRepo
	suggestedEmailsRepo repo.SuggestedEmailRepo
	authService         security.AuthService
}

func NewUserService(db *sqlx.DB,
	userRepo repo.UserRepo,
	deviceRepo repo.DeviceRepo,
	planRepo repo.PlanRepo,
	suggestedEmailsRepo repo.SuggestedEmailRepo,
	authService security.AuthService,
) UserService {
	return &userService{
		userRepo:            userRepo,
		deviceRepo:          deviceRepo,
		planRepo:            planRepo,
		suggestedEmailsRepo: suggestedEmailsRepo,
		authService:         authService,
	}
}

func (s *userService) Create(device Device) *CreatedUser {
	var jwt string
	var userId uuid.UUID
	var deviceId uuid.UUID
	var err error

	txFn := func(tx *sqlx.Tx) error {
		userId = s.userRepo.Create(tx)
		s.deviceRepo.DeleteByFingerprint(tx, device.Fingerprint)
		device.UserID = userId
		deviceId = s.deviceRepo.Create(tx, device)
		if jwt, err = s.authService.CreateToken(userId, deviceId); err != nil {
			return err
		}
		return nil
	}

	if err = dbs.WithTx(txFn); err != nil {
		panic(err)
	}

	return &CreatedUser{ID: userId, DeviceID: deviceId, Jwt: jwt}
}

func (s *userService) SendMeOtp(email string) string {
	var verifySid string
	if slices.Contains(configs.TestEmails, email) {
		verifySid = configs.TestSID
	} else {
		var err error
		verifySid, err = emails.SendOtp(email)
		if err != nil {
			panic(err)
		}
	}
	return verifySid
}

func (s *userService) VerifyOtp(meta Meta, email, sid, otp string) *VerifiedUser {
	var otpStatus string
	var err error

	if slices.Contains(configs.TestEmails, email) && sid == configs.TestSID && otp == configs.TestOTP {
		otpStatus = "approved"
	} else {
		otpStatus, err = emails.VerifyOtp(otp, sid, email)
	}

	if otpStatus != "approved" {
		panic("OTP not verified for " + email + ", status: " + otpStatus)
	}

	var jwt string
	var newUserId UUID
	var user *User

	txFn := func(tx *sqlx.Tx) error {
		user = s.userRepo.GetOneByEmail(email)
		if user == nil {
			s.userRepo.UpdateEmail(tx, meta.UserID, email)
			newUserId = meta.UserID
			logs.Info(uuid.Nil, "User loggedIn for %s", email)
		} else {
			s.planRepo.UpdateUserID(tx, meta.UserID, user.ID)
			devices := s.deviceRepo.GetMany(user.ID)

			if len(devices) >= 5 {
				s.deviceRepo.Delete(tx, devices[len(devices)-1].ID)
			}
			s.deviceRepo.UpdateUserID(tx, meta.DeviceID, user.ID)
			s.userRepo.Delete(tx, meta.UserID)
			newUserId = user.ID
			logs.Info(uuid.Nil, "Merging userId:%s to %s", meta.UserID, user.ID)
		}

		jwt, err = s.authService.CreateToken(newUserId, meta.DeviceID)
		return err
	}

	if err = dbs.WithTx(txFn); err != nil {
		panic(err)
	}

	userFullName := ""
	if user != nil && user.Name != nil {
		userFullName = *user.Name
	}
	return &VerifiedUser{
		UserID:       newUserId,
		DeviceID:     meta.DeviceID,
		Jwt:          jwt,
		UserFullName: &userFullName,
		Email:        &email,
	}
}

func (s *userService) RefreshToken(meta Meta) *VerifiedUser {
	user := s.userRepo.GetOne(meta.UserID)
	jwt, err := s.authService.CreateToken(meta.UserID, meta.DeviceID)
	if err != nil {
		return nil
	}
	return &VerifiedUser{
		UserID:       meta.UserID,
		DeviceID:     meta.DeviceID,
		Jwt:          jwt,
		UserFullName: user.Name,
		Email:        user.Email,
	}
}

func (s *userService) UpdateName(userID UUID, name string) int64 {
	return s.userRepo.UpdateName(userID, name)
}

func (s *userService) Logout(userID UUID, deviceId UUID) int64 {

	device := s.deviceRepo.GetOne(deviceId)
	if device.UserID != userID {
		panic("invalid deviceId")
	}

	var rows int64 = 0
	dbs.WithTx(func(tx *sqlx.Tx) error {
		rows = s.deviceRepo.Delete(tx, deviceId)
		return nil
	})
	return int64(rows)
}

func (s *userService) Delete(userID UUID, sid, otp string) {
	user := s.userRepo.GetOne(userID)
	var otpStatus string
	var err error
	if slices.Contains(configs.TestEmails, *user.Email) && sid == configs.TestSID && otp == configs.TestOTP {
		otpStatus = "approved"
	} else {
		otpStatus, err = emails.VerifyOtp(otp, sid, *user.Email)
		if err != nil {
			panic("OTP not verified for " + *user.Email + ", status: " + otpStatus)
		}
	}

	if otpStatus != "approved" {
		panic("OTP not verified for " + *user.Email + ", status: " + otpStatus)
	}

	dbs.WithTx(func(tx *sqlx.Tx) error {
		s.suggestedEmailsRepo.DeleteManyByEmail(*user.Email)
		s.userRepo.Delete(tx, userID)
		return nil
	})

}

func (s *userService) GetDevices(userID UUID) []Device {
	return s.deviceRepo.GetMany(userID)
}

func (s *userService) GetSuggestedEmails(userID UUID) []SuggestedEmail {
	return s.suggestedEmailsRepo.GetMany(userID)
}

func (s *userService) DeleteSuggestedEmail(userID UUID, suggestedEmailId UUID) {
	suggestedEmail := s.suggestedEmailsRepo.GetOne(suggestedEmailId)
	if suggestedEmail == nil || suggestedEmail.UserID != userID {
		panic("invalid suggestedEmailId")
	}
	s.suggestedEmailsRepo.Delete(suggestedEmailId)
}

File: mahaam-api-cs/Src/Feat/User/UserService.cs
using System.Transactions;
using Mahaam.Feat.Plans;
using Mahaam.Infra;

namespace Mahaam.Feat.Users;

public interface IUserService
{
	Task<CreatedUser> Create(Device device);
	Task<string?> SendMeOtp(string email);
	Task<VerifiedUser> VerifyOtp(string email, string sid, string otp);
	Task<VerifiedUser> RefreshToken();
	Task UpdateName(string name);
	Task Logout(Guid deviceId);
	Task Delete(string sid, string otp);
	Task<List<Device>> GetDevices();
	Task<List<SuggestedEmail>> GetSuggestedEmails();
	Task DeleteSuggestedEmail(Guid suggestedEmailId);
}

class UserService(IUserRepo userRepo, IDeviceRepo deviceRepo, IPlanRepo planRepo, ISuggestedEmailsRepo suggestedEmailsRepo, ILog log, IAuth auth, IEmail emailService, Settings settings) : IUserService
{
	public async Task<CreatedUser> Create(Device device)
	{
		using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
		var userId = await userRepo.Create();

		// add device
		device.UserId = userId;
		await deviceRepo.DeleteByFingerprint(device.Fingerprint);
		var deviceId = await deviceRepo.Create(device);

		string jwt = auth.CreateToken(userId!.ToString(), deviceId.ToString());
		scope.Complete();

		log.Info($"User Created with id:{userId}, deviceId:{device.Id}.");
		return new CreatedUser { Id = userId, DeviceId = deviceId, Jwt = jwt };
	}

	public async Task<string?> SendMeOtp(string email)
	{
		string? verifySid;
		if (settings.Email.TestEmails.Contains(email)) verifySid = settings.Email.TestSID;
		else verifySid = emailService.SendOtp(email);

		if (verifySid != null) log.Info($"OTP sent to {email}");

		return verifySid;
	}

	public async Task<VerifiedUser> VerifyOtp(string email, string sid, string otp)
	{
		string otpStatus;
		if (settings.Email.TestEmails.Contains(email) && sid.Equals(settings.Email.TestSID) && otp.Equals(settings.Email.TestOTP))
			otpStatus = "approved";
		else
			otpStatus = emailService.VerifyOtp(otp, sid, email);

		if (!"approved".Equals(otpStatus))
			throw new ArgumentException($"OTP not verified for {email}, status: {otpStatus}");

		using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
		var user = await userRepo.GetOne(email);
		var deviceId = Req.DeviceId;
		if (user is null)
		{
			await userRepo.UpdateEmail(Req.UserId, email);
			log.Info($"User loggedIn for {email}");
		}
		else
		{
			// move plans of current user to the one with email
			await planRepo.UpdateUserId(Req.UserId, user.Id);
			var devices = await deviceRepo.GetMany(user.Id);
			if (devices != null && devices.Count >= 5)
			{
				await deviceRepo.Delete(devices.Last().Id);
			}

			await deviceRepo.UpdateUserId(deviceId, user.Id);
			await userRepo.Delete(Req.UserId);
			log.Info($"Merging userId:{Req.UserId} to {user.Id}");
		}


		var newUserId = user is null ? Req.UserId! : user.Id;
		string jwt = auth.CreateToken(newUserId.ToString(), deviceId.ToString());
		scope.Complete();

		log.Info($"OTP verified for {email}");
		return new VerifiedUser { UserId = newUserId, DeviceId = deviceId, Jwt = jwt, UserFullName = user?.Name, Email = email };
	}

	public async Task<VerifiedUser> RefreshToken()
	{
		var user = await userRepo.GetOne(Req.UserId);
		string jwt = auth.CreateToken(Req.UserId.ToString(), Req.DeviceId.ToString());

		return new VerifiedUser { UserId = Req.UserId, DeviceId = Req.DeviceId, Jwt = jwt, UserFullName = user?.Name, Email = user?.Email };
	}

	public async Task UpdateName(string name)
	{
		await userRepo.UpdateName(Req.UserId, name);
	}

	public async Task Logout(Guid deviceId)
	{
		var device = await deviceRepo.GetOne(deviceId);
		if (device is null || !device.UserId.Equals(Req.UserId))
			throw new UnauthorizedException("Invalid deviceId");
		await deviceRepo.Delete(deviceId);
	}

	public async Task DeleteSuggestedEmail(Guid suggestedEmailId)
	{
		var suggestedEmail = await suggestedEmailsRepo.GetOne(suggestedEmailId);
		if (suggestedEmail is null || !suggestedEmail.UserId.Equals(Req.UserId))
			throw new UnauthorizedException("Invalid suggestedEmailId");
		await suggestedEmailsRepo.Delete(suggestedEmailId);
	}

	public async Task Delete(string sid, string otp)
	{
		var user = await userRepo.GetOne(Req.UserId);

		string otpStatus;
		if (user.Email != null && settings.Email.TestEmails.Contains(user.Email) && sid.Equals(settings.Email.TestSID) && otp.Equals(settings.Email.TestOTP))
			otpStatus = "approved";
		else
			otpStatus = emailService.VerifyOtp(otp, sid, user.Email ?? string.Empty);

		if (!"approved".Equals(otpStatus))
			throw new ArgumentException($"OTP not approved for {user.Email ?? "unknown"}, status: {otpStatus}");

		using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);

		if (user.Email != null)
			await suggestedEmailsRepo.DeleteManyByEmail(user.Email);
		await userRepo.Delete(Req.UserId);
		scope.Complete();
	}

	public async Task<List<Device>> GetDevices()
	{
		return await deviceRepo.GetMany(Req.UserId);
	}

	public async Task<List<SuggestedEmail>> GetSuggestedEmails()
	{
		return await suggestedEmailsRepo.GetMany(Req.UserId);
	}
}